documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,232 @@
+
+* Use async mode (write-behind) if possible
+* Prevent cache misses by preloading data
+* For JDBC Cache Store:
+** Use indexes on `id` column to prevent table scans
+** Use PRIMARY_KEY on `id` column
+** Configure batch-size, fetch-size, etc
+
+== Hints for program developers
+
+There are also several hints for developers which can be easily applied to the client application and will boost up the performance.
+
+=== Ignore return values
+
+When you're not interested in returning value by `#put` or `#remove` method, use `Flag.IGNORE_RETURN_VALUES` flag as shown below:
+
+[source,java]
+.Flag.IGNORE_RETURN_VALUES
+----
+cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES)
+----
+
+=== Skip cache load
+
+When you don't want to load an entry from a Cache Store, use `Flag.SKIP_CACHE_LOAD` as shown below:
+
+[source,java]
+.Flag.SKIP_CACHE_LOAD
+----
+cache.getAdvancedCache().withFlags(Flag.SKIP_CACHE_LOAD)
+----
+
+=== Use `Externalizable` for marshalling
+
+Infinispan uses JBoss Marshalling to transfer objects over the wire. The most efficient way to marshall user data is to provide an Externalizer:
+
+[source,java]
+.User entity with Externalizer
+----
+import org.infinispan.marshall.AdvancedExternalizer;
+
+public class Book {
+
+   final String name;
+   final String author;
+
+   public Book(String name, String author) {
+      this.name = name;
+      this.author = author;
+   }
+
+   public static class BookExternalizer
+            implements AdvancedExternalizer<Book> {
+
+      @Override
+      public void writeObject(ObjectOutput output, Book book)
+            throws IOException {
+         output.writeObject(book.name);
+         output.writeObject(book.author);
+      }
+
+      @Override
+      public Person readObject(ObjectInput input)
+            throws IOException, ClassNotFoundException {
+         return new Person((String) input.readObject(), (String) input.readObject());
+      }
+
+      @Override
+      public Set<Class<? extends Book>> getTypeClasses() {
+         return Util.<Class<? extends Book>>asSet(Book.class);
+      }
+
+      @Override
+      public Integer getId() {
+         return 2345;
+      }
+   }
+}
+----
+
+The Externalizer must be registered in cache configuration. See configuration examples below:
+
+[source,java]
+.Adding Externalizer using XML
+----
+<cache-container>
+   <serialization>
+      <advanced-externalizer class="Book$BookExternalizer"/>
+   </serialization>
+</cache-container>
+----
+
+[source,java]
+.Adding Externalizer using Java
+----
+GlobalConfigurationBuilder builder = ...
+builder.serialization().addAdvancedExternalizer(new Book.BookExternalizer());
+----
+
+=== Storing Strings efficiently
+
+When the majority of stored data are Strings there are additional optimizations which can be taken into consideration:
+
+* Using `String#getBytes("UTF-8")` allows to decrease size of the object
```

Comment:
I'd rephrase this as "If your strings are mostly ASCII, convert them to `UTF-8` and store them as `byte[]`. Note that the `UTF-8` version can be bigger for non-ASCII strings."